src/browser.js

```diff
@@ -40,38 +41,47 @@ class UniversalEmitter extends Emitter {
   emit(type: mixed, payload: mixed): void {
     payload = super.mapEvent(type, payload);
     super.handleEvent(type, payload);
-    this.batch.push({type, payload});
+    this.storage.add({type, payload});
   }
   // match server api
   from(): UniversalEmitter {
     return this;
   }
-  flushInternal(): void {
-    if (this.batch.length > 0) {
-      this.fetch('/_events', {
+  async flushInternal(): Promise<void> {
+    const items = this.storage.getAndClear();
+    if (items.length === 0) return;
+
+    try {
+      const res = await this.fetch('/_events', {
         method: 'POST',
         headers: {
           'Content-Type': 'application/json',
         },
-        body: JSON.stringify({items: this.batch} || []),
-      }).catch(() => {});
+        body: JSON.stringify({items}),
+      });
+
+      if (!res.ok) {
+        // sending failed so put the logs back into storage
+        this.storage.addToStart(...items);
+      }
+    } catch (e) {
+      // sending failed so put the logs back into storage
+      this.storage.addToStart(...items);
     }
-    this.batch = [];
   }
   teardown(): void {
     window.removeEventListener('beforeunload', this.flush);
     clearInterval(this.interval);
     this.interval = null;
-    this.batch = [];
   }
 }
 
 const plugin =
   __BROWSER__ &&
   createPlugin({
-    deps: {fetch: FetchToken},
-    provides: ({fetch}) => {
-      return new UniversalEmitter(fetch);
+    deps: {fetch: FetchToken, storage: UniversalEventsBatchStorageToken},
```

Comment:
Could we instead make this Optional and default to localStorage to prevent a breaking change?

Reply:
I didn't realize you could do that, will update!

Reply:
fixed